bitswap/network/ipfs_impl.go

```diff
@@ -108,29 +135,39 @@ func (s *streamMessageSender) Connect(ctx context.Context) (network.Stream, erro
 		return nil, err
 	}
 
-	s.stream = stream
-	s.connected = true
-	return s.stream, nil
+	s.stream.Store(stream)
+	return stream, nil
 }
 
 // Reset the stream
 func (s *streamMessageSender) Reset() error {
-	if s.stream != nil {
-		err := s.stream.Reset()
-		s.connected = false
+	stream := s.stream.Load()
+	if stream != nil {
+		err := stream.Reset()
+		s.stream.Store(nil)
 		return err
 	}
 	return nil
 }
 
 // Close the stream
 func (s *streamMessageSender) Close() error {
-	return s.stream.Close()
```

Comment:
wouldn't nil-ifying `s.stream` here also free the stream reference to be GC'ed ?

Reply:
Also, I don't think `Close()` is used. And Reset() only on queue shutdown. Even or read/write errors the stream isn't closed nor reset, it seems.

Reply:
Maybe we should add
```go
    defer sender.Close()
```
at this [location](https://github.com/ipfs/boxo/blob/main/bitswap/client/internal/messagequeue/messagequeue.go#L558).

Reply:
(the location doesn't make sense?)

I guess there's nothing wrong with re-using a stream and only cleaning up on peer disconnect?

Reply:
> wouldn't nil-ifying s.stream here also free the stream reference to be GC'ed ?

Are you suggesting we don't close the stream here and drop our reference? That would be wrong. We should clean up our resources when we are done with them.

Ideally we'd use one stream per message and close the stream once we're done as @gammazero suggests. I believe the reason we don't do that is that in Yamux the initial send buffer is [constant](https://github.com/hashicorp/yamux/blob/060257c97bcf6317b24df54b9f8abe8e451d4606/const.go#L125) per stream (rather than [configurable](https://datatracker.ietf.org/doc/html/rfc9000#section-4) both within a stream and per connection connection as in QUIC). This means that if our initial message is larger than that initial send capacity, then we have to wait for a round trip to send the whole message. By reusing the stream, the peer can increase our send buffer size in a later message.

It may be worth changing this in Yamux.

Reply:
> I guess there's nothing wrong with re-using a stream and only cleaning up on peer disconnect?

The real issue in this code is that the cleanup of peer disconnects is slow. A lot of time is spent on acquiring the lock by many different goroutines (ping me on slack and I'll point you to the pprofs). This PR is merely a small refactor in opportunistically dropping the stream reference if we know it's done.

The benefit of the one stream per message approach is that you avoid the whole issue of maintaining a reference and knowing if a connection is even alive or not.

Reply:
I really like the idea of having one stream per message. Reusing the stream is like double-muxing just to avoid a buffer sizing issue. Being able to have different messages on the same connection is point of muxing and streams is the mechanism for doing that.

The initial stream buffer size is 256K. If most of our messages are smaller, then that is fine. I suspect we do need a larger buffer, particularly for block transfer. Seems like a change to yamux is worth it, unless the occasional extra round trip is not a huge problem.

Reply:
Discussion moved to issue: https://github.com/ipfs/boxo/issues/765